Fix17857

bendk opened the pull request.

Commit:
bz17857 part 2: make mouse scroll work on the progress slider.

On GTK, we make it so one click goes 5 seconds.  On OSX respond to scrollWheel
method.

If you can use the mouse wheel, then you may not need to use the slider at
all.

Commit:
bz17857 part 1:  Implement slider_size() correctly.

This makes it so when the user clicks on the progress slider, we center the
position of the slider on the pointer.  Before you basically couldn't know
where the slider would end up.

Commit:
bz17857 part 3: Don't move the progress slider if the user clicks on it

This makes dragging feel more natural.  If the user clicks slightly off
center, we shouldn't seek.

Added the get_slider_pos() method to CustomSlider and started using it in the
portable code.

paulswartz commented on tv/lib/frontends/widgets/gtk/customcontrols.py:
```diff
@@ -159,32 +154,78 @@ def gtk_scale_class(self):
+        if pos_difference > self.slider_size() / 2.0:
+            self.move_slider(event_pos)
+        self.drag_info = _DragInfo(event.button,
+                                   self.get_slider_pos(),
```
Can be `current_pos`?

bendk replied:
current_pos is the position at the start of the event.  If we move the slider, we need to call get_slider_pos() again.

paulswartz replied:
On OS X, that's done by resetting `current_pos = self.get_slider_pos()` and using the variable.  That avoids calling `get_slider_pos()` twice in some of the cases.

paulswartz commented on tv/lib/frontends/widgets/gtk/customcontrols.py:
```diff
@@ -376,14 +429,38 @@ def set_value(self, value):
         # Try to pick a reasonable default for the digits
         range = max_value - min_value
-        self._widget.set_digits(int(round(math.log10(100.0 / range))))
+        self._widget.set_digits(int(round(math.log10(10000.0 / range))))
```
What does this magic # do?

bendk replied:
I'm not 100% sure, but I think it sets the precision of the slider based on the number of digits.  Before we would use 2 digits of precision, but that made using the scrollwheel weird, since we could be seeking by less than 0.1.  After the change we get 4 digits of precision.

paulswartz replied:
Can you put that in a comment?

paulswartz commented on tv/osx/plat/frontends/widgets/customcontrol.py:
```diff
@@ -198,6 +200,9 @@ def sendAction_to_(self, action, to):
         return YES
 DragableDrawableButton.setCellClass_(DragableButtonCell)
 
+MouseTrackingInfo = collections.namedtuple("MouseTrackingInfo",
```
Is there a reason this is implemented differently from `_DragInfo` on GTK?

bendk replied:
I think there might be slight differences across platforms, but mostly I just didn't take the time figure out how to put the code in portable.

paulswartz commented on tv/osx/plat/frontends/widgets/customcontrol.py:
```diff
@@ -251,6 +294,25 @@ def isOpaque(self):
+        range = self.maxValue() - self.minValue()
+        value_change = (event.deltaY() / size) * range
+        self.setFloatValue_(self.floatValue() + value_change)
+        wrapper.emit('changed', self.floatValue())
```
Why don't we emit the 'pressed' and 'released' signals here?

bendk replied:
oops, I'll add that.

paulswartz commented on tv/osx/plat/frontends/widgets/customcontrol.py:
```diff
@@ -352,7 +417,7 @@ def set_range(self, min_value, max_value):
 
-    def set_increments(self, increment, big_increment):
+    def set_increments(self, small_step, big_step, scroll_step=None):
         pass
```
The documentation from line 301 should probably be here as well.

bendk replied:
I thought that we decided that GTK was the official place for docstrings for platform-specific code.  The reason not to put it in both is then when we change things, we need to update it in 2 places and probably won't.

paulswartz replied:
I mean this:

``` python
# NOTE: we ignore the scroll_step value passed into set_increments()
# and calculate the change using deltaY, which is in device
# coordinates.
```

It is platform-specific.  It doesn't need to be a docstring; just a note about why we ignore the values.

bendk replied:
Oops, totally misread that one.  Fixed now.

Commit:
Improvements based on paul's review.

Commit:
More improvements.

Added comments to set_increments on OSX.

Commit:
Avoid calling get_slider_pos() so much.

I changed this based on pauls comments and then realized we actually don't
have to call get_slider_pos() again in any case.
